CONTRIBUTING.md

```diff
@@ -0,0 +1,602 @@
+        3.  macro that modifies the meaning or adds functionality to the
+            previous / following statement.
+
+        However, simply calling function `f` from function `g` does not imply
+        that `f` and `g` are "related".
+2.  Statements within structures / classes are separated with 1 blank line.
+    There are no blank lines betweeen the first / last statement in the
+    structure / class.
+    1.  _exception_: there is no blank line between an access modifier (`private`, `protected`, `public`) and the following statement.
+       _example_:
+        ```c++
+        class foo {
+        public:
+            int get_x() const noexcept { return x_; }
+
+            int &get_x() noexcept { return x_; }
+
+        private:
+            int x_;
+        };
+        ```
+
+3.  Function bodies cannot have multiple consecutive blank lines, and a single
+    blank line can only appear between two logical sections of the function.
+4. Unit tests should follow the [AAA](http://wiki.c2.com/?ArrangeActAssert)
+   pattern, and a single blank line must appear between consecutive "A"
+   sections. No other blank lines are allowed in unit tests.
+5.  Enumeration definitions should have no blank lines between consecutive
+    enumerators.
+
+
+### Include statement grouping
+
+In general, all include statements should be present on the top of the file,
+ordered in the following groups, with two blank lines between each group:
+
+1. Related header file (e.g. `core/foo/bar.hpp` included in `core/foo/bar.cpp`,
+   or in the unit test`core/test/foo/bar.cpp`)
+2. Standard library headers (e.g. `vector`)
+3. Executor specific library headers (e.g. `omp.h`)
+4. System third-party library headers (e.g. `papi.h`)
+5. Local third-party library headers
+6. Public Ginkgo headers
+7. Private Ginkgo headers
+
+_Example_: A file `core/base/my_file.cpp` might have an include list like this:
+
+```c++
+#include <ginkgo/core/base/my_file.hpp>
+
+
+#include <algorithm>
+#include <vector>
+#include <tuple>
+
+
+#include <omp.h>
+
+
+#include <papi.h>
+
+
+#include "third_party/blas/cblas.hpp"
+#include "third_party/lapack/lapack.hpp"
+
+
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+
+
+#include "core/base/my_file_kernels.hpp"
+```
+
+#### Main header
+
+This section presents general rules used to define the main header attributed to
+the file. In the previous example, this would be ` #include
+<ginkgo/core/base/my_file.hpp>`.
+
+General rules:
+1. Some fixed main header.
+2. components:
+  - with `_kernel` suffix looks for the header in the same folder.
+  - without `_kernel` suffix looks for the header in `core`.
+3. `test/utils`: looks for the header in `core`
+4. `core`: looks for the header in `ginkgo`
+5. `test` or `base`: looks for the header in `ginkgo/core`
+6. others: looks for the header in `core`
+
+_Note_: Please see the detail in the `dev_tools/scripts/config`.
+
```

Comment:
```suggestion

#### Some general comments.

1. Private headers of Ginkgo should not be included within the public Ginkgo header.
2. It is a good idea of keep the headers self-sufficient, See [Google Style guide for reasoning](https://google.github.io/styleguide/cppguide.html#Self_contained_Headers). 
3. The recommendations of the `iwyu` (Include what you use) tool can be used to make sure that the headers are self-sufficient and that the compiled files ( `.cu`, '.cpp', `.hip.cpp` ) include only what they use. A [CI pipeline](https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/jobs/584358356) is available that runs with the `iwyu` tool. Please be aware that this tool can be incorrect in some cases.
```

Reply:
@tcojean , how about here ?

Reply:
point 2., `it is a good idea **to** keep the`...

Otherwise, seems good to me ! Thanks.